R/sampler-adaptive.R

```diff
@@ -0,0 +1,310 @@
+    } else {
+      pars_remove <- NULL
+      internal$weight <- internal$weight + 1 
+    }
+    
+    internal$scaling <- 
+      update_scaling(internal$scaling, internal$iteration, accept_prob,
+                     internal$scaling_increment, min_scaling, acceptance_target, 
+                     pre_diminish, internal$n_start, log_scaling_update)
+    internal$scaling_history <- c(internal$scaling_history, internal$scaling)
+    internal$autocorrelation <- update_autocorrelation(
+      state$pars, internal$weight, internal$autocorrelation, pars_remove)
+    internal$mean <- update_mean(state$pars, internal$weight, internal$mean,
+                                 pars_remove)
+    internal$vcv <- update_vcv(internal$mean, internal$autocorrelation,
+                               internal$weight)
+    internal$included <- 
+      update_included(internal$included, internal$iteration, is_replacement)
+
+    state
+  }
+
+  ## Marc; do we want all of these saved out?  Did we ever decide how
+  ## to merge these over multiple chains?
+  finalise <- function(state, model, rng) {
+    as.list(internal)
+  }
+
+  mcstate_sampler("Adaptive Metropolis-Hastings",
+                  initialise,
+                  step,
+                  finalise)
+}
+
+
+calc_scaling_increment <- function(d, acceptance_target, log_scaling_update) {
+  if (log_scaling_update) {
+    A <- - qnorm(acceptance_target / 2)
+    
+    scaling_increment <- 
+      (1 - 1 / d) * (sqrt(2 * pi) * exp(A ^ 2 / 2)) / (2 * A) + 
+      1 / (d * acceptance_target * (1 - acceptance_target))
+  } else {
+    scaling_increment <- 1 / 100
+  }
+  
+  scaling_increment
+}
+
+
+calc_n_start <- function(acceptance_target) {
+  5 / (acceptance_target * (1 - acceptance_target))
+}
+
+
+qp <- function(x) {
+  outer(x, x)
+}
+
+
+calc_proposal_vcv <- function(scaling, vcv, weight, initial_vcv,
+                              initial_vcv_weight) {
+  d <- dim(vcv)[1]
+  
+  weighted_vcv <-
+    ((weight - 1) * vcv + (initial_vcv_weight + d + 1) * initial_vcv) /
+    (weight + initial_vcv_weight + d + 1)
+  
+  2.38 ^ 2 / d * scaling ^ 2 * weighted_vcv
+}
+
+
+check_replacement <- function(iteration, forget_rate, forget_end) {
+  is_forget_step <- floor(forget_rate * iteration) >
+    floor(forget_rate * (iteration - 1))
+  is_before_forget_end <- iteration <= forget_end
+  
+  is_forget_step & is_before_forget_end
+}
+
+
+update_scaling <- function(scaling, iteration, accept_prob, scaling_increment,
+                           min_scaling, acceptance_target, pre_diminish, 
+                           n_start, log_scaling_update) {
+  scaling_change <- scaling_increment * (accept_prob - acceptance_target) /
+    sqrt(n_start + max(0, iteration - pre_diminish))
+  
+  if (log_scaling_update) {
+    scaling <- pmax(min_scaling, scaling * exp(scaling_change))
+  } else {
+    scaling <- pmax(min_scaling, scaling + scaling_change)
+  }
+  
```

Comment:
```suggestion
  if (log_scaling_update) {
    pmax(min_scaling, scaling * exp(scaling_change))
  } else {
    pmax(min_scaling, scaling + scaling_change)
  }
```

but are these vectors here (min_scaling, scaling and scaling_change)?